src/services/textChanges.ts

```diff
@@ -590,104 +586,82 @@ namespace ts.textChanges {
             });
         }
 
-        public getChanges(): FileTextChanges[] {
+        /**
+         * Note: after calling this, the TextChanges object must be discarded!
+         * @param validate only for tests
+         *    The reason we must validate as part of this method is that `getNonFormattedText` changes the node's positions,
+         *    so we can only call this once and can't get the non-formatted text separately.
+         */
+        public getChanges(validate?: ValidateNonFormattedText): FileTextChanges[] {
             this.finishInsertNodeAtClassStart();
-            return group(this.changes, c => c.sourceFile.path).map(changesInFile => {
+            return changesToText.changesToText(this.changes, this.newLineCharacter, this.formatContext, validate);
+        }
+    }
+
+    export type ValidateNonFormattedText = (node: Node, text: string) => void;
+
+    namespace changesToText {
+        export function changesToText(changes: ReadonlyArray<Change>, newLineCharacter: string, formatContext: formatting.FormatContext, validate: ValidateNonFormattedText): FileTextChanges[] {
+            return group(changes, c => c.sourceFile.path).map(changesInFile => {
                 const sourceFile = changesInFile[0].sourceFile;
-                const textChanges = ChangeTracker.normalize(changesInFile).map(c =>
-                    createTextChange(createTextSpanFromRange(c.range), this.computeNewText(c, sourceFile)));
+                // order changes by start position
+                const normalized = stableSort(changesInFile, (a, b) => a.range.pos - b.range.pos);
+                // verify that change intervals do not overlap, except possibly at end points.
+                for (let i = 0; i < normalized.length - 2; i++) {
+                    Debug.assert(normalized[i].range.end <= normalized[i + 1].range.pos);

```

Comment:
I realize this code was copied, but I don't think this assert would be actionable if it fired.

Reply:
We could fix the codefix/refactoring that was adding overlapping changes?

Reply:
My experience has been that asserts in loops are hard to address because there's no information about where in the loop the assert occurred.  If all we need to know is that some range was bad and the rest of the stack will tell you the area in which the problem occurred, then this is fine.